### PR TITLE
Support clobber_abi and vector registers (clobber-only) in PowerPC inline assembly

### PR DESCRIPTION
**Pull request by taiki-e**
This supports `clobber_abi` which is one of the requirements of stabilization mentioned in #93335.

This basically does a similar thing I did in https://github.com/rust-lang/rust/pull/130630 to implement `clobber_abi` for s390x, but for powerpc/powerpc64/powerpc64le.
- This also supports vector registers (as `vreg`) as clobber-only, which need to support clobbering of them to implement `clobber_abi`.
- `vreg` should be able to accept `#[repr(simd)]` types as input/output if the unstable `altivec` target feature is enabled, but `core::arch::{powerpc,powerpc64}` vector types, `#[repr(simd)]`, and `core::simd` are all unstable, so the fact that this is currently a clobber-only should not be considered a blocker of clobber_abi implementation or stabilization. So I have not implemented it in this PR.
  - See https://github.com/rust-lang/rust/pull/131551 (which is based on this PR) for a PR to implement this.
  - (I'm not sticking to whether that PR should be a separate PR or part of this PR, so I can merge that PR into this PR if needed.)

Refs:
- PPC32 SysV: Section "Function Calling Sequence" in [System V Application Binary Interface PowerPC Processor Supplement](https://refspecs.linuxfoundation.org/elf/elfspec_ppc.pdf)
  EDIT: this document is old; See section 3.2 "Function Calling Sequence" in [Power Architecture 32-bit Application Binary Interface Supplement 1.0 - Linux & Embedded](https://web.archive.org/web/20120608163804/https://www.power.org/resources/downloads/Power-Arch-32-bit-ABI-supp-1.0-Unified.pdf) instead. (https://github.com/rust-lang/rust/pull/132638)
- PPC64 ELFv1: Section 3.2 "Function Calling Sequence" in [64-bit PowerPC ELF Application Binary Interface Supplement](https://refspecs.linuxfoundation.org/ELF/ppc64/PPC-elf64abi.html#FUNC-CALL)
- PPC64 ELFv2: Section 2.2 "Function Calling Sequence" in [64-Bit ELF V2 ABI Specification](https://openpowerfoundation.org/specifications/64bitelfabi/)
- AIX: [Register usage and conventions](https://www.ibm.com/docs/en/aix/7.3?topic=overview-register-usage-conventions), [Special registers in the PowerPC®](https://www.ibm.com/docs/en/aix/7.3?topic=overview-special-registers-in-powerpc), [AIX vector programming](https://www.ibm.com/docs/en/aix/7.3?topic=concepts-aix-vector-programming)
- Register definition in LLVM: https://github.com/llvm/llvm-project/blob/llvmorg-19.1.0/llvm/lib/Target/PowerPC/PPCRegisterInfo.td#L189

If I understand the above four ABI documentations correctly, except for the PPC32 SysV's VR (Vector Registers) and 32-bit AIX (currently not supported by rustc)'s r13, there does not appear to be important differences in terms of implementing `clobber_abi`:
- The above four ABIs are consistent about FPR (0-13: volatile, 14-31: nonvolatile), CR (0-1,5-7: volatile, 2-4: nonvolatile), XER (volatile), and CTR (volatile).
- As for GPR, only the registers we are treating as reserved are slightly different
  - r0, r3-r12 are volatile
  - r1(sp, reserved), r14-31 are nonvolatile
  - r2(reserved) is TOC pointer in PPC64 ELF/AIX, system-reserved register in PPC32 SysV (AFAIK used as thread pointer in Linux/BSDs)
  - r13(reserved for non-32-bit-AIX) is thread pointer in PPC64 ELF, small data area pointer register in PPC32 SysV, "reserved under 64-bit environment; not restored across system calls[^r13]" in AIX)
- As for FPSCR, volatile in PPC64 ELFv1/AIX, some fields are volatile only in certain situations (rest are volatile) in PPC32 SysV/PPC64 ELFv2.
- As for VR (Vector Registers), it is not mentioned in PPC32 SysV, v0-v19 are volatile in both in PPC64 ELF/AIX, v20-v31 are nonvolatile in PPC64 ELF, reserved or nonvolatile depending on the ABI ([vec-extabi vs vec-default in LLVM](https://reviews.llvm.org/D89684), we are [using vec-extabi](https://github.com/rust-lang/rust/pull/131341#discussion_r1797693299)) in AIX:
  > When the default Vector enabled mode is used, these registers are reserved and must not be used.
  > In the extended ABI vector enabled mode, these registers are nonvolatile and their values are preserved across function calls

  I left [FIXME comment about PPC32 SysV](https://github.com/rust-lang/rust/pull/131341#discussion_r1790496095) and added ABI check for AIX.
- As for VRSAVE, it is not mentioned in PPC32 SysV, nonvolatile in PPC64 ELFv1, reserved in PPC64 ELFv2/AIX
- As for VSCR, it is not mentioned in PPC32 SysV/PPC64 ELFv1, some fields are volatile only in certain situations (rest are volatile) in PPC64 ELFv2, volatile in AIX

We are currently treating r1-r2, r13 (non-32-bit-AIX), r29-r31, LR, CTR, and VRSAVE as reserved.
We are currently not processing anything about FPSCR and VSCR, but I feel those are things that should be processed by `preserves_flags` rather than `clobber_abi` if we need to do something about them. (However, PPCRegisterInfo.td in LLVM does not seem to define anything about them.)

Replaces #111335 and #124279

cc @ecnelises @bzEq @lu-zero

r? @Amanieu

@rustbot label +O-PowerPC +A-inline-assembly

[^r13]: callee-saved, according to [LLVM](https://github.com/llvm/llvm-project/blob/6a6af0246bd2d68291582e9aefc0543e5c6102fe/llvm/lib/Target/PowerPC/PPCCallingConv.td#L322) and [GCC](https://github.com/gcc-mirror/gcc/blob/a9173a50e7e346a218323916e4d3add8552529ae/gcc/config/rs6000/rs6000.h#L859).

### CODE REVIEW

File: compiler/rustc_target/src/asm/mod.rs
Comment (taiki-e) [inline compiler/rustc_target/src/asm/mod.rs:1127]: The reason I did not treat these as clobbered only on PPC64 is that the PPC32 ABI document I referenced was released at a time when Altivec/VMX did not exist, and I thought it might not reflect the final status of the PPC32 ABI.

In a similar case, the early ABI documents for s390x (e.g., the one mentioned [here](https://github.com/rust-lang/rust/pull/111331#issuecomment-1537788377)) do not mention vector registers, but the ABI documents since the addition of vector facility mention them and all are treated as volatile.

File: compiler/rustc_target/src/asm/powerpc.rs
Comment (taiki-e) [inline compiler/rustc_target/src/asm/powerpc.rs:139]: @ecnelises @bzEq: Is the default in Rust the default Vector enabled mode, as the name implies? Also, is there a way provided for the compiler to understand the current mode?

(If the first is yes and the second is no, it would be sufficient to simply reject the use of v20-v31 as reserved. If the first is no, this code is fine as is.)

**Comment by taiki-e** [comment]
> - `vreg` should be able to accept `#[repr(simd)]` types as input/output if the unstable `altivec` target feature is enabled, but `core::arch::{powerpc,powerpc64}` vector types, `#[repr(simd)]`, and `core::simd` are all unstable, so the fact that this is currently a clobber-only should not be considered a blocker of clobber_abi implementation or stabilization. So I have not implemented it in this PR.

Opened https://github.com/rust-lang/rust/pull/131551 (which is based on this PR) to implement this.

**Comment by bzEq** [comment]
> r13 is thread pointer in PPC64 ELF, ..., "reserved under 64-bit environment" in AIX

In fact, `r13` is also thread pointer in AIX-64 pthread environment, can be considered reserved.

File: compiler/rustc_target/src/asm/powerpc.rs
Comment (bzEq) [reply compiler/rustc_target/src/asm/powerpc.rs:139]: It should refer to `vec-extabi`, which is the default. See https://github.com/rust-lang/rust/blob/8f8bee4f60d9d3769f75c70d558c27a95761c554/compiler/rustc_target/src/spec/base/aix.rs#L6.

**Comment by taiki-e** [comment]
Updated to check ABI in v20-v31 for AIX, and update def_regs and clobber_abi to reflect the fact that r13 is not reserved and "not restored across system calls" on 32-bit AIX (see also [LLVM's getReservedRegs](https://github.com/llvm/llvm-project/blob/6a6af0246bd2d68291582e9aefc0543e5c6102fe/llvm/lib/Target/PowerPC/PPCRegisterInfo.cpp#L354-L440)).

File: src/doc/unstable-book/src/language-features/asm-experimental-arch.md
Comment (bzEq) [inline src/doc/unstable-book/src/language-features/asm-experimental-arch.md:163]: It's ok to not support these two registers conservasively IMO, might rephrase to `These might be used as base pointer inside LLVM`.

File: src/doc/unstable-book/src/language-features/asm-experimental-arch.md
Comment (bzEq) [inline src/doc/unstable-book/src/language-features/asm-experimental-arch.md:162]: nit: Separate `r2` and `r13`.
`r2` is used as TOC pointer.
`r13` is system reserved register.

File: src/doc/unstable-book/src/language-features/asm-experimental-arch.md
Comment (bzEq) [inline src/doc/unstable-book/src/language-features/asm-experimental-arch.md:164]: I think `xer` should be not supported either and should occupy a line after `lr`.

Comment (taiki-e) [reply src/doc/unstable-book/src/language-features/asm-experimental-arch.md:164]: Currently, `xer` is supported as a clobber-only register.

https://github.com/rust-lang/rust/blob/d30a53e46962b7bd6bcc74e23cfb81c36e996eed/src/doc/unstable-book/src/language-features/asm-experimental-arch.md?plain=1#L39

See https://rust-lang.zulipchat.com/#narrow/stream/216763-project-inline-asm/topic/preserve_flags.20on.20OpenPower.2FPowerPC.3F/near/251292056 for more context.

File: src/doc/unstable-book/src/language-features/asm-experimental-arch.md
Comment (taiki-e) [reply src/doc/unstable-book/src/language-features/asm-experimental-arch.md:162]: Those explanations were just copied from the error message, but AFAIK r2 is used as a thread pointer in the PPC32 SVR4 ABI ([Linux](https://github.com/bminor/musl/blob/v1.2.5/arch/powerpc/pthread_arch.h), [FreeBSD](https://github.com/freebsd/freebsd-src/blob/release/14.1.0/sys/powerpc/include/tls.h#L52), [NetBSD](https://github.com/NetBSD/src/blob/432a1357026b10c184d8a0ddb683008a23cc7cd9/sys/arch/powerpc/include/mcontext.h#L159), [OpenBSD](https://github.com/openbsd/src/blob/ed8f5e8d82ace15e4cefca2c82941b15cb1a7830/sys/arch/powerpc/include/tcb.h#L37)), so I think the current explanation is actually fine as is.

https://github.com/rust-lang/rust/blob/d30a53e46962b7bd6bcc74e23cfb81c36e996eed/compiler/rustc_target/src/asm/powerpc.rs#L197-L198

File: src/doc/unstable-book/src/language-features/asm-experimental-arch.md
Comment (taiki-e) [reply src/doc/unstable-book/src/language-features/asm-experimental-arch.md]: I found a line that collectively describes registers in the same situation in other architectures, so I moved these two there.

https://github.com/rust-lang/rust/blob/ca59dfec2902a08f5921f5f16f6aa5c741466d0e/src/doc/unstable-book/src/language-features/asm-experimental-arch.md?plain=1#L155

File: src/doc/unstable-book/src/language-features/asm-experimental-arch.md
Comment (bzEq) [reply src/doc/unstable-book/src/language-features/asm-experimental-arch.md:163]: Cool, I thought these lines are for input or output operands. It makes sense `xer` as clobber operand since there are instructions changing the `ca` bit of `xer`.

**Comment by bzEq** [comment]
I have no more comments and LGTM.

**Review by bzEq — APPROVED** [review]

File: compiler/rustc_target/src/asm/powerpc.rs
Comment (workingjubilee) [inline compiler/rustc_target/src/asm/powerpc.rs:85]: Noting for myself: I _believe_ this should genuinely not be reachable from user code that isn't using target-spec.json, but I am not 100% confident about this. I should at some point refactor targets so they _can_ emit real errors in these cases, instead. I do not consider this a fact to block this PR over, however, as there is a lot of kinda-questionable handling of some things that _might_ be user-reachable for _all_ the targets, and the worst case here is just an ICE.

File: compiler/rustc_target/src/asm/mod.rs
Comment (workingjubilee) [inline compiler/rustc_target/src/asm/mod.rs:948]: Noting for myself: I find it slightly amusing that "sysv64" is not valid on platforms which... er... use a System V 64-bit ABI... but it "makes sense" because it's for the AMD64 ABI specifically on Windows-like platforms.

File: src/doc/unstable-book/src/language-features/asm-experimental-arch.md
Comment (workingjubilee) [inline src/doc/unstable-book/src/language-features/asm-experimental-arch.md:165]: Is there a UI test for these? akin to https://github.com/rust-lang/rust/blob/4d88de2acd2cf8595c61e5c8d39183e5579204d8/tests/ui/asm/x86_64/bad-reg.rs

Comment (taiki-e) [reply src/doc/unstable-book/src/language-features/asm-experimental-arch.md:165]: Added ui test for this: https://github.com/rust-lang/rust/pull/131341/files#diff-ad23a8c300c887c5ca754ccb89f97d9a7c5cf714df4ba260adb0055b6678f456

Btw, such unsupported registers are present in most architectures, but only aarch64, x86_64, and not yet merged [sparc/sparc64](https://github.com/rust-lang/rust/pull/132472/files#diff-02aebda3376c2b020265137f9ce2c387669ca5cfecd7d60494275c2387db5114) (and this PR) currently have ui tests for them.  I plan to add tests for other arches later.

**Comment by workingjubilee** [comment]
Thanks!

@bors r=bzEq,workingjubilee

File: compiler/rustc_target/src/asm/mod.rs
Comment (taiki-e) [reply compiler/rustc_target/src/asm/mod.rs:1145]: According to [Power Architecture 32-bit Application Binary Interface Supplement 1.0 - Linux & Embedded](https://web.archive.org/web/20120608163804/https://www.power.org/resources/downloads/Power-Arch-32-bit-ABI-supp-1.0-Unified.pdf), PPC32 has the same convention here as PPC64.

Therefore, we can just remove the FIXME comment here.

**Pull request merged**
